File: gtci/01-two-pointers/cpp/3-squaring-sorted-array.cpp
#include <cassert>
#include <cmath>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
   public:
    static vector<int> makeSquares(const vector<int>& arr) {
        int n = arr.size();
        vector<int> squares(n);

        std::vector<int>::iterator squaresItr = squares.end() - 1;

        std::vector<int>::const_iterator left = arr.begin();
        std::vector<int>::const_iterator right = arr.end() - 1;

        while (left <= right) {
            if (abs(*left) > abs(*right)) {
                *squaresItr = *left * *left;
                squaresItr--;
                left++;
            } else if (abs(*right) >= abs(*left)) {
                *squaresItr = *right * *right;
                squaresItr--;
                right--;
            }
        }

        return squares;
    }
};

int main() {
    Solution sol;
    std::vector<int> arr = {-2, -1, 0, 1, 2};
    std::vector<int> result = sol.makeSquares(arr);
    std::vector<int> expected1 = {0, 1, 1, 4, 4};
    assert(result == expected1);

    std::vector<int> arr2 = {-3, -2, -1, 0, 1, 2, 3, 3};
    std::vector<int> result2 = sol.makeSquares(arr2);
    std::vector<int> expected2 = {0, 1, 1, 4, 4, 9, 9, 9};
    assert(result2 == expected2);

    std::vector<int> arr3 = {-3, -2, -1};
    std::vector<int> result3 = sol.makeSquares(arr3);
    std::vector<int> expected3 = {1, 4, 9};
    assert(result3 == expected3);

    std::cout << "All test cases passed." << std::endl;
    return 0;
}

File: gtci/01-two-pointers/python3/3-squaring-sorted-array.py
class Solution:
    def makeSquares(self, arr):
        n = len(arr)
        squares = [0 for x in range(n)]

        squares_ptr = n - 1

        left = 0
        right = n - 1

        while left <= right:
            if abs(arr[left]) > abs(arr[right]):
                squares[squares_ptr] = arr[left] ** 2
                squares_ptr -= 1
                left += 1
            elif abs(arr[right]) >= abs(arr[left]):
                squares[squares_ptr] = arr[right] ** 2
                squares_ptr -= 1
                right -= 1

        return squares


if __name__ == "__main__":
    sol = Solution()
    arr = [-2, -1, 0, 1, 2]
    result = sol.makeSquares(arr)
    assert result == [0, 1, 1, 4, 4]

    arr2 = [-3, -2, -1, 0, 1, 2, 3, 3]
    result2 = sol.makeSquares(arr2)
    assert result2 == [0, 1, 1, 4, 4, 9, 9, 9]

    arr3 = [-3, -2, -1]
    result3 = sol.makeSquares(arr3)
    assert result3 == [1, 4, 9]
    print("All test cases passed.")
